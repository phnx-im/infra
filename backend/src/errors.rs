// SPDX-FileCopyrightText: 2024 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use thiserror::Error;
use tracing::error;

#[derive(Debug, Error)]
pub enum StorageError {
    #[error(transparent)]
    Database(#[from] DatabaseError),
    #[error("Error deserializing column: {0}")]
    Serde(#[from] phnxtypes::codec::Error),
}

impl From<sqlx::Error> for StorageError {
    fn from(e: sqlx::Error) -> Self {
        Self::Database(e.into())
    }
}

impl From<Box<dyn std::error::Error + Send + Sync>> for StorageError {
    fn from(e: Box<dyn std::error::Error + Send + Sync>) -> Self {
        Self::Database(e.into())
    }
}

impl From<StorageError> for tonic::Status {
    fn from(e: StorageError) -> Self {
        match e {
            StorageError::Database(error) => {
                error!(%error, "Database error");
                Self::internal("Database error")
            }
            StorageError::Serde(error) => {
                error!(%error, "Seriazation error");
                Self::internal("Seriazation error")
            }
        }
    }
}

#[derive(Debug, Error)]
pub enum DatabaseError {
    #[error(transparent)]
    Sqlx(#[from] sqlx::Error),
    #[error(transparent)]
    Dynamic(#[from] Box<dyn std::error::Error + Send + Sync>),
}

/// General error while accessing the requested queue.
#[derive(Error, Debug)]
pub(super) enum QueueError {
    #[error(transparent)]
    Storage(#[from] StorageError),
    /// Mismatching sequence numbers.
    #[error("Mismatching sequence numbers.")]
    SequenceNumberMismatch,
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
}

impl From<sqlx::Error> for QueueError {
    fn from(e: sqlx::Error) -> Self {
        Self::Storage(e.into())
    }
}

impl From<phnxtypes::codec::Error> for QueueError {
    fn from(e: phnxtypes::codec::Error) -> Self {
        Self::Storage(e.into())
    }
}
