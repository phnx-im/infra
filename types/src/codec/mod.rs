// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use cbor::Cbor;
use error::CodecError;
use mls_assist::memory_provider::Codec;
use serde::{Serialize, de::DeserializeOwned};

mod cbor;
mod error;
mod persistence;
#[cfg(test)]
mod tests;

pub use error::Error;
pub use persistence::{BlobDecoded, BlobEncoded};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
#[repr(u8)]
#[non_exhaustive]
pub enum PhnxCodec {
    #[cfg(test)]
    OlderTestVersion = 0,
    #[default]
    V1 = 1,
}

impl TryFrom<u8> for PhnxCodec {
    type Error = Error;

    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            #[cfg(test)]
            0 => Ok(PhnxCodec::OlderTestVersion),
            1 => Ok(PhnxCodec::V1),
            _ => Err(Error::UnknownCodecVersion),
        }
    }
}

impl PhnxCodec {
    fn serialize_to_writer<T: Serialize>(
        &self,
        value: &T,
        writer: &mut impl std::io::Write,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
        // The first byte is always the codec version
        writer.write_all(&[*self as u8])?;
        match self {
            #[cfg(test)]
            PhnxCodec::OlderTestVersion => tests::Json::to_writer(value, writer)?,
            PhnxCodec::V1 => Cbor::to_writer(value, writer)?,
        }
        Ok(())
    }

    fn serialize<T: Sized + Serialize>(
        &self,
        value: &T,
    ) -> Result<Vec<u8>, Box<dyn std::error::Error + Send + Sync + 'static>> {
        let mut buf = Vec::new();
        self.serialize_to_writer(value, &mut buf)?;
        Ok(buf)
    }

    fn deserialize<T: DeserializeOwned>(
        &self,
        bytes: &[u8],
    ) -> Result<T, Box<dyn std::error::Error + Send + Sync + 'static>> {
        let res = match self {
            #[cfg(test)]
            PhnxCodec::OlderTestVersion => tests::Json::from_slice(bytes)?,
            PhnxCodec::V1 => Cbor::from_slice(bytes)?,
        };
        Ok(res)
    }

    pub fn to_vec<T>(value: &T) -> Result<Vec<u8>, Error>
    where
        T: Sized + Serialize,
    {
        let codec_version = PhnxCodec::default();
        let res = codec_version.serialize(value).map_err(|error| CodecError {
            codec_version,
            error,
        })?;
        Ok(res)
    }

    pub fn from_slice<T>(bytes: &[u8]) -> Result<T, Error>
    where
        T: DeserializeOwned,
    {
        let codec_version_byte = bytes.first().ok_or(Error::EmptyyInputSlice)?;
        let codec_version = PhnxCodec::try_from(*codec_version_byte)?;
        codec_version.deserialize(&bytes[1..]).map_err(|error| {
            CodecError {
                codec_version,
                error,
            }
            .into()
        })
    }
}

impl Codec for PhnxCodec {
    type Error = Error;

    fn to_vec<T>(value: &T) -> Result<Vec<u8>, Self::Error>
    where
        T: Sized + Serialize,
    {
        Self::to_vec(value)
    }

    fn from_slice<T>(bytes: &[u8]) -> Result<T, Self::Error>
    where
        T: DeserializeOwned,
    {
        Self::from_slice(bytes)
    }
}
