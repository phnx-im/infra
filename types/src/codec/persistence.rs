// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use serde::{Serialize, de::DeserializeOwned};
use sqlx::{Database, Decode, Encode, Type, encode::IsNull, error::BoxDynError};

use super::PhnxCodec;

pub struct BlobEncoded<T: Serialize>(pub T);

impl<DB: Database, T: Serialize> Type<DB> for BlobEncoded<T>
where
    Vec<u8>: Type<DB>,
{
    fn type_info() -> <DB as Database>::TypeInfo {
        Vec::<u8>::type_info()
    }
}

impl<'q, DB: Database, T: Serialize> Encode<'q, DB> for BlobEncoded<T>
where
    Vec<u8>: Encode<'q, DB>,
{
    fn encode_by_ref(
        &self,
        buf: &mut <DB as Database>::ArgumentBuffer<'q>,
    ) -> Result<IsNull, BoxDynError> {
        let bytes = PhnxCodec::to_vec(&self.0)?;
        Encode::<DB>::encode(&bytes, buf)
    }
}

#[derive(Debug)]
pub struct BlobDecoded<T>(pub T);

impl<T> BlobDecoded<T> {
    pub fn into_inner(self) -> T {
        self.0
    }
}

impl<DB: Database, T: DeserializeOwned> Type<DB> for BlobDecoded<T>
where
    Vec<u8>: Type<DB>,
{
    fn type_info() -> <DB as Database>::TypeInfo {
        Vec::<u8>::type_info()
    }
}

impl<'q, DB: Database, T: DeserializeOwned> Decode<'q, DB> for BlobDecoded<T>
where
    for<'a> &'a [u8]: Decode<'a, DB>,
{
    fn decode(value: <DB as Database>::ValueRef<'q>) -> Result<Self, BoxDynError> {
        let bytes: &[u8] = Decode::<DB>::decode(value)?;
        PhnxCodec::from_slice(bytes)
            .map(BlobDecoded)
            .map_err(From::from)
    }
}
