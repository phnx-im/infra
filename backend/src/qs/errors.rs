// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use crate::errors::{QueueError, StorageError};

use super::network_provider::NetworkProvider;
use aircommon::crypto::errors::{DecryptionError, EncryptionError, KeyGenerationError};
use thiserror::Error;
use tracing::error;

// === DS API errors ===

/// Error fetching a message from the QS.
#[derive(Error, Debug)]
pub enum QsEnqueueError<N: NetworkProvider> {
    /// Unseal error
    #[error(transparent)]
    UnsealError(#[from] DecryptionError),
    /// An error ocurred enqueueing in a fan out queue
    #[error(transparent)]
    EnqueueError(#[from] EnqueueError),
    /// An error ocurred while sending a message to the network
    #[error("An error ocurred while sending a message to the network")]
    NetworkError(N::NetworkError),
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid response
    #[error("Invalid response")]
    InvalidResponse,
}

/// Error enqueuing a fanned-out message.
#[derive(Error, Debug)]
pub enum EnqueueError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError, // E.g. an error while encoding a message before enqueing it.
    /// Error in the underlying storage provider
    #[error("Error in the underlying storage provider")]
    Storage(StorageError),
    /// Client not found
    #[error("Client not found")]
    ClientNotFound,
    /// Queue error
    #[error("Queue error")]
    Queue(QueueError),
}

impl From<StorageError> for EnqueueError {
    fn from(error: StorageError) -> Self {
        error!(%error, "Failed to enqueue message due to storage error");
        EnqueueError::Storage(error)
    }
}

impl From<sqlx::Error> for EnqueueError {
    fn from(error: sqlx::Error) -> Self {
        error!(%error, "Failed to enqueue message due to sqlx error");
        EnqueueError::Storage(error.into())
    }
}

impl From<QueueError> for EnqueueError {
    fn from(error: QueueError) -> Self {
        error!(%error, "Failed to enqueue message due to queue error");
        match error {
            QueueError::Storage(storage_error) => storage_error.into(),
            QueueError::SequenceNumberMismatch | QueueError::PayloadReceiverClosed => {
                Self::Queue(error)
            }
        }
    }
}

impl From<EncryptionError> for EnqueueError {
    fn from(error: EncryptionError) -> Self {
        error!(%error, "Failed to enqueue message due to encryption error");
        EnqueueError::LibraryError
    }
}

// === Internal errors ===

#[derive(Debug, Error)]
pub(super) enum GenerateAndStoreError {
    #[error("Error generating signature keypair")]
    KeyGenerationError(#[from] KeyGenerationError),
    #[error("Error storing key")]
    StorageError(#[from] StorageError),
}
