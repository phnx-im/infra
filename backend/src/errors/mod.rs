// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use displaydoc::Display;
use mls_assist::{
    group, memory_provider::MlsAssistMemoryStorage, openmls::group::MergeCommitError,
};
use thiserror::Error;
use tokio::sync::mpsc;
use tonic::Status;
use tracing::error;

use aircommon::codec::PersistenceCodec;

pub(crate) mod auth_service;
pub(crate) mod qs;

pub(crate) type CborMlsAssistStorage = MlsAssistMemoryStorage<PersistenceCodec>;

#[derive(Debug, Error)]
pub enum StorageError {
    #[error(transparent)]
    Database(#[from] DatabaseError),
    #[error("Error deserializing column: {0}")]
    Serde(#[from] aircommon::codec::Error),
}

impl From<sqlx::Error> for StorageError {
    fn from(e: sqlx::Error) -> Self {
        Self::Database(e.into())
    }
}

impl From<Box<dyn std::error::Error + Send + Sync>> for StorageError {
    fn from(e: Box<dyn std::error::Error + Send + Sync>) -> Self {
        Self::Database(e.into())
    }
}

impl From<StorageError> for Status {
    fn from(error: StorageError) -> Self {
        error!(%error, "storage error");
        match error {
            StorageError::Database(_) => Self::internal("Database error"),
            StorageError::Serde(_) => Self::internal("Seriazation error"),
        }
    }
}
#[derive(Debug, Error)]
pub enum DatabaseError {
    #[error(transparent)]
    Sqlx(#[from] sqlx::Error),
    #[error(transparent)]
    Dynamic(#[from] Box<dyn std::error::Error + Send + Sync>),
}

/// General error while accessing the requested queue.
#[derive(Error, Debug, Display)]
pub(super) enum QueueError {
    /// {0:?}
    Storage(#[from] StorageError),
    /// Mismatching sequence numbers
    SequenceNumberMismatch,
    /// Payload receiver closed
    PayloadReceiverClosed,
}

impl From<sqlx::Error> for QueueError {
    fn from(e: sqlx::Error) -> Self {
        Self::Storage(e.into())
    }
}

impl From<aircommon::codec::Error> for QueueError {
    fn from(e: aircommon::codec::Error) -> Self {
        Self::Storage(e.into())
    }
}

impl<T> From<mpsc::error::SendError<T>> for QueueError {
    fn from(_: mpsc::error::SendError<T>) -> Self {
        Self::PayloadReceiverClosed
    }
}

impl From<QueueError> for Status {
    fn from(error: QueueError) -> Self {
        let msg = error.to_string();
        match error {
            QueueError::Storage(error) => {
                error!(%error, "storage error");
                Self::internal(msg)
            }
            QueueError::SequenceNumberMismatch => Self::internal(msg),
            QueueError::PayloadReceiverClosed => Self::internal(msg),
        }
    }
}

/// Potential errors when performing a group operation.
#[derive(Debug, Error)]
pub(crate) enum GroupOperationError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message")]
    ProcessingError,
    /// Missing queue config in client key package.
    #[error("Missing queue config in client key package")]
    MissingQueueConfig,
    /// Incomplete Welcome message.
    #[error("Incomplete Welcome message.")]
    IncompleteWelcome,
    #[error("Error merging commit")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<GroupOperationError> for Status {
    fn from(e: GroupOperationError) -> Self {
        let msg = e.to_string();
        match e {
            GroupOperationError::LibraryError
            | GroupOperationError::ProcessingError
            | GroupOperationError::InvalidMessage => Status::internal(msg),
            GroupOperationError::MissingQueueConfig | GroupOperationError::IncompleteWelcome => {
                Status::invalid_argument(msg)
            }
            GroupOperationError::MergeCommitError(merge_commit_error) => {
                error!(%merge_commit_error, "failed merging commit");
                Status::internal(msg)
            }
        }
    }
}

/// Potential errors when joining a connection group.
#[derive(Debug, Error)]
pub(crate) enum JoinConnectionGroupError {
    /// Invalid assisted message.
    #[error("Invalid assisted message")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message")]
    ProcessingError,
    /// Not a connection group.
    #[error("Not a connection group")]
    NotAConnectionGroup,
    #[error("Error merging commit")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<JoinConnectionGroupError> for Status {
    fn from(e: JoinConnectionGroupError) -> Self {
        let msg = e.to_string();
        match e {
            JoinConnectionGroupError::InvalidMessage
            | JoinConnectionGroupError::NotAConnectionGroup => Status::invalid_argument(msg),
            JoinConnectionGroupError::ProcessingError => Status::internal(msg),
            JoinConnectionGroupError::MergeCommitError(merge_commit_error) => {
                error!(%merge_commit_error, "failed merging commit");
                Status::internal(msg)
            }
        }
    }
}

/// Potential errors when deleting a group.
#[derive(Debug, Error)]
pub(crate) enum GroupDeletionError {
    /// Invalid assisted message.
    #[error("Invalid assisted message")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message")]
    ProcessingError,
    #[error("Error merging commit")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<GroupDeletionError> for Status {
    fn from(e: GroupDeletionError) -> Self {
        let msg = e.to_string();
        match e {
            GroupDeletionError::InvalidMessage => Status::invalid_argument(msg),
            GroupDeletionError::ProcessingError => Status::internal(msg),
            GroupDeletionError::MergeCommitError(merge_commit_error) => {
                error!(%merge_commit_error, "failed merging commit");
                Status::internal(msg)
            }
        }
    }
}

/// Potential errors when processing a self remove proposal.
#[derive(Debug, Error)]
pub(crate) enum ClientSelfRemovalError {
    /// Invalid assisted message.
    #[error("Invalid assisted message")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message")]
    ProcessingError,
    #[error("Error merging commit")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<ClientSelfRemovalError> for Status {
    fn from(e: ClientSelfRemovalError) -> Self {
        let msg = e.to_string();
        match e {
            ClientSelfRemovalError::InvalidMessage => Status::invalid_argument(msg),
            ClientSelfRemovalError::ProcessingError => Status::internal(msg),
            ClientSelfRemovalError::MergeCommitError(merge_commit_error) => {
                error!(%merge_commit_error, "failed merging commit");
                Status::internal(msg)
            }
        }
    }
}

/// Potential errors when resyncing a client.
#[derive(Debug, Error)]
pub(crate) enum ResyncClientError {
    /// Invalid assisted message.
    #[error("Invalid assisted message")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message")]
    ProcessingError,
    #[error("Error merging commit")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<ResyncClientError> for Status {
    fn from(e: ResyncClientError) -> Self {
        let msg = e.to_string();
        match e {
            ResyncClientError::InvalidMessage => Status::invalid_argument(msg),
            ResyncClientError::ProcessingError => Status::internal(msg),
            ResyncClientError::MergeCommitError(merge_commit_error) => {
                error!(%merge_commit_error, "failed merging commit");
                Status::internal(msg)
            }
        }
    }
}
