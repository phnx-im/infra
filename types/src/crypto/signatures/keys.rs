// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use mls_assist::{openmls::prelude::SignaturePublicKey, openmls_rust_crypto::OpenMlsRustCrypto};
use serde::{Deserialize, Serialize};
use tls_codec::{TlsDeserializeBytes, TlsSerialize, TlsSize};

use crate::crypto::errors::KeyGenerationError;

use super::{
    private_keys::{SigningKey, VerifyingKey},
    traits::{SigningKeyBehaviour, VerifyingKeyBehaviour},
};

#[derive(Debug)]
pub struct LeafVerifyingKey(VerifyingKey);

impl VerifyingKeyBehaviour for LeafVerifyingKey {}

impl AsRef<VerifyingKey> for LeafVerifyingKey {
    fn as_ref(&self) -> &VerifyingKey {
        &self.0
    }
}

impl From<&SignaturePublicKey> for LeafVerifyingKey {
    fn from(pk_ref: &SignaturePublicKey) -> Self {
        Self(pk_ref.clone().into())
    }
}

#[derive(
    Clone,
    PartialEq,
    Eq,
    Serialize,
    Deserialize,
    Debug,
    TlsSerialize,
    TlsDeserializeBytes,
    TlsSize,
    sqlx::Type,
)]
#[sqlx(transparent)]
pub struct QsClientVerifyingKey(VerifyingKey);

impl QsClientVerifyingKey {
    #[cfg(any(test, feature = "test_utils"))]
    pub fn new_for_test(verifying_key: VerifyingKey) -> Self {
        Self(verifying_key)
    }
}

impl AsRef<VerifyingKey> for QsClientVerifyingKey {
    fn as_ref(&self) -> &VerifyingKey {
        &self.0
    }
}

impl VerifyingKeyBehaviour for QsClientVerifyingKey {}

#[derive(Clone, Serialize, Deserialize)]
pub struct QsClientSigningKey(SigningKey);

impl QsClientSigningKey {
    pub fn random() -> Result<Self, KeyGenerationError> {
        let rust_crypto = OpenMlsRustCrypto::default();
        let signing_key = SigningKey::generate()?;
        Ok(Self(signing_key))
    }

    pub fn verifying_key(&self) -> QsClientVerifyingKey {
        QsClientVerifyingKey(self.0.verifying_key().clone())
    }
}

impl AsRef<SigningKey> for QsClientSigningKey {
    fn as_ref(&self) -> &SigningKey {
        &self.0
    }
}

impl super::traits::SigningKeyBehaviour for QsClientSigningKey {}

#[derive(
    Clone,
    PartialEq,
    Serialize,
    Deserialize,
    Debug,
    TlsSerialize,
    TlsDeserializeBytes,
    TlsSize,
    sqlx::Type,
)]
#[sqlx(transparent)]
pub struct QsUserVerifyingKey(VerifyingKey);

impl QsUserVerifyingKey {
    #[cfg(any(test, feature = "test_utils"))]
    pub fn new_for_test(verifying_key: VerifyingKey) -> Self {
        Self(verifying_key)
    }
}

impl AsRef<VerifyingKey> for QsUserVerifyingKey {
    fn as_ref(&self) -> &VerifyingKey {
        &self.0
    }
}

impl VerifyingKeyBehaviour for QsUserVerifyingKey {}

#[derive(Clone, Serialize, Deserialize)]
pub struct QsUserSigningKey(SigningKey);

impl QsUserSigningKey {
    pub fn generate() -> Result<Self, KeyGenerationError> {
        let signing_key = SigningKey::generate()?;
        Ok(Self(signing_key))
    }

    pub fn verifying_key(&self) -> QsUserVerifyingKey {
        QsUserVerifyingKey(self.0.verifying_key().clone())
    }
}

impl AsRef<SigningKey> for QsUserSigningKey {
    fn as_ref(&self) -> &SigningKey {
        &self.0
    }
}

impl SigningKeyBehaviour for QsUserSigningKey {}

#[cfg(test)]
mod test {
    use crate::codec::PhnxCodec;

    use super::*;

    #[test]
    fn qs_client_verifying_key_serde_codec() {
        let key = QsClientVerifyingKey::new_for_test(VerifyingKey::new_for_test(vec![1, 2, 3]));
        insta::assert_binary_snapshot!(".cbor", PhnxCodec::to_vec(&key).unwrap());
    }

    #[test]
    fn qs_client_verifying_key_serde_json() {
        let key = QsClientVerifyingKey::new_for_test(VerifyingKey::new_for_test(vec![1, 2, 3]));
        insta::assert_json_snapshot!(key);
    }
}
