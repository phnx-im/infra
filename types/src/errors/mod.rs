// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use mls_assist::{
    group::errors::StorageError, memory_provider::MlsAssistMemoryStorage,
    openmls::group::MergeCommitError,
};
use thiserror::Error;
use tracing::error;
use version::VersionError;

use crate::codec::PhnxCodec;

pub mod auth_service;
pub mod qs;
pub mod version;

pub type CborMlsAssistStorage = MlsAssistMemoryStorage<PhnxCodec>;

/// Potential errors when performing a group operation.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum GroupOperationError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    /// Missing queue config in client key package.
    #[error("Missing queue config in client key package.")]
    MissingQueueConfig,
    /// Failed to retrieve QS verifying key.
    #[error("Failed to retrieve QS verifying key.")]
    FailedToObtainVerifyingKey,
    /// User added twice.
    #[error("User added twice.")]
    DuplicatedUserAddition,
    /// Incomplete Welcome message.
    #[error("Incomplete Welcome message.")]
    IncompleteWelcome,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when updating a client.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum ClientUpdateError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    /// Unknown sender.
    #[error("Unknown sender.")]
    UnknownSender,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when processing a message.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum DsProcessingError {
    /// API Version error
    #[error(transparent)]
    Api(#[from] VersionError),
    /// Failed to distribute message to other members
    #[error("Failed to distribute message to other members")]
    DistributionError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Invalid signature.
    #[error("Invalid signature.")]
    InvalidSignature,
    /// Group not found.
    #[error("Group not found.")]
    GroupNotFound,
    /// Could not decrypt group state.
    #[error("Could not decrypt group state.")]
    CouldNotDecrypt,
    /// Could not encrypt group state.
    #[error("Could not decrypt group state.")]
    CouldNotEncrypt,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    /// Unknown sender.
    #[error("Unknown sender.")]
    UnknownSender,
    /// Invalid sender type.
    #[error("Invalid sender type")]
    InvalidSenderType,
    /// Error storing encrypted group state.
    #[error("Error storing encrypted group state.")]
    StorageError,
    /// Error creating group.
    #[error("Failed to create group: Group ID not reserved")]
    UnreservedGroupId,
    /// Error updating client.
    #[error(transparent)]
    ClientUpdateError(#[from] ClientUpdateError),
    /// Could not find welcome info for this sender and/or this epoch.
    #[error("Could not find welcome info for this sender and/or this epoch.")]
    NoWelcomeInfoFound,
    /// Error joining group.
    #[error(transparent)]
    JoinGroupError(#[from] JoinGroupError),
    /// Error joining connection group.
    #[error(transparent)]
    JoinConnectionGroupError(#[from] JoinConnectionGroupError),
    /// Error adding clients.
    #[error(transparent)]
    ClientAddtionError(#[from] ClientAdditionError),
    /// Error removing clients.
    #[error(transparent)]
    ClientRemovalError(#[from] ClientRemovalError),
    /// Error resyncing client.
    #[error(transparent)]
    ClientResyncError(#[from] ResyncClientError),
    /// Error self removing client.
    #[error(transparent)]
    ClientSelfRemovalError(#[from] ClientSelfRemovalError),
    /// Error deleting group.
    #[error(transparent)]
    GroupDeletionError(#[from] GroupDeletionError),
    /// Error performing group operation.
    #[error(transparent)]
    GroupOperationError(#[from] GroupOperationError),
    /// Deprecated parameter
    #[error("Deprecated param: {0}")]
    DeprecatedParam(&'static str),
}

/// Potential errors when joining a group.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum JoinGroupError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    /// Unknown sender.
    #[error("Unknown sender.")]
    UnknownSender,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when joining a connection group.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum JoinConnectionGroupError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    /// Unknown sender.
    #[error("Unknown sender.")]
    UnknownSender,
    /// Not a connection group.
    #[error("Not a connection group.")]
    NotAConnectionGroup,
    /// User auth key collision.
    #[error("User auth key collision.")]
    UserAuthKeyCollision,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when adding a user.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum ClientAdditionError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    /// Missing queue config in client key package.
    #[error("Missing queue config in client key package.")]
    MissingQueueConfig,
    /// Incomplete Welcome message.
    #[error("Incomplete Welcome message.")]
    IncompleteWelcome,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when removing clients.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum ClientRemovalError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when deleting a group.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum GroupDeletionError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when processing a self remove proposal.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum ClientSelfRemovalError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when sending a message.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum MessageSendingError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
}

/// Potential errors when resyncing a client.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum ResyncClientError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when validating a commit or proposal.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum ValidationError {
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
}
