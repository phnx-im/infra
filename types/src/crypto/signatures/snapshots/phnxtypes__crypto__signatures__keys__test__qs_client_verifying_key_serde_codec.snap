---
source: types/src/crypto/signatures/keys.rs
expression: "PhnxCodec::to_vec(&key).unwrap()"
extension: cbor
snapshot_kind: binary
---
